plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -1779,6 +1779,11 @@ producer_plugin_impl::start_block_result producer_plugin_impl::start_block() {
    if (!chain_plug->accept_transactions())
       return start_block_result::waiting_for_block;
 
+   abort_block();
+
+   chain.maybe_switch_forks([this](const transaction_metadata_ptr& trx) { _unapplied_transactions.add_forked(trx); },
+                            [this](const transaction_id_type& id) { return _unapplied_transactions.get_trx(id); });
```

Comment:
Do we still need this?

Reply:
I think we do once #2159 is done. #2159 will put blocks into the fork database that could potentially be better forks. So a check here will allow us to switch to that fork if needed while producing.

Reply:
It makes sense that we have to do this, but I would have expected this deciding what is the best fork to be in `controller.cpp`, maybe in `start_block`. However I'm not very familiar with `producer_plugin`, or the architectural reason why it is separated into a plugin and not just in `libraries/chain`,  so there is probably a good reason why this has to be where you added it.

Reply:
Yeah, has to be here. We need to switch forks before determining who should produce which is determined in the `producer_plugin`. I have often said that `producer_plugin` should be in the chain library. However, I now know of a use case where it is really nice that it is a plugin. It allows for different producing logic while still using the core controller.